File: server/python/controllers/transaction.py
from flask import jsonify, make_response
from .fetch_graphql import braintree_fetch_graphql


def braintree_graphql_transaction(transactionData):
    try:
        payment_token = transactionData["paymentToken"]

        query = """
        mutation ($input: ChargeCreditCardInput!) {
            chargeCreditCard(input: $input) {
                transaction {
                    id
                    status
                }
            }
        }
        """

        variables = {
            "input": {
                "paymentMethodId": payment_token["id"],
                "transaction": {
                    "amount": "1.00",
                    "riskData": {"deviceData": transactionData["deviceData"]},
                    "shipping": {
                        "shippingAddress": transactionData["shippingAddress"],
                        "shippingMethod": "GROUND",
                    },
                    "customerDetails": {
                        "email": transactionData["email"],
                    },
                    "vaultPaymentMethodAfterTransacting": {
                        "when": "ON_SUCCESSFUL_TRANSACTION"
                    },
                },
                "options": {
                    "billingAddress": payment_token["paymentSource"]["card"][
                        "billingAddress"
                    ]
                },
            }
        }

        payload = {"query": query, "variables": variables}
        response = braintree_fetch_graphql(payload)

        if "errors" in response:
            errors = response["errors"]
            print(errors)
            return jsonify({"error": errors}), 500

        result = response["data"]["chargeCreditCard"]["transaction"]
        return jsonify({"result": result})

    except Exception as error:
        print(error)
        return make_response(jsonify(error=str(error)), 500)

File: server/node/src/controllers/transaction.js
import { fetchGraphql } from '../lib/fetch-graphql.js';

export async function createTransaction(req, res) {
  try {
    const { email, deviceData, paymentToken, shippingAddress } = req.body;

    const query = `
      mutation ($input: ChargeCreditCardInput!) {
        chargeCreditCard(input: $input) {
          transaction {
            id
            legacyId
            createdAt
            amount {
              value
              currencyCode
            }
            status
          }
        }
      }
    `;
    const variables = {
      input: {
        paymentMethodId: paymentToken.id,
        transaction: {
          amount: '1.00',
          riskData: { deviceData },
          shipping: {
            shippingAddress,
            shippingMethod: 'GROUND',
          },
          customerDetails: { email },
          vaultPaymentMethodAfterTransacting: {
            when: 'ON_SUCCESSFUL_TRANSACTION',
          },
        },
        options: {
          billingAddress: paymentToken.paymentSource.card.billingAddress,
        },
      },
    };
    const response = await fetchGraphql({ query, variables });
    const { data, errors } = await response.json();

    if (errors) {
      console.error(errors);
      res.status(500).json({ error: errors });
      return;
    }
    res.json({ result: data.chargeCreditCard.transaction });
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: error.message });
  }
}
